src/Uno.UI/UI/Xaml/UIElement.cs

```diff
@@ -39,194 +39,247 @@ public partial class UIElement : DependencyObject
+
+		internal void RaiseEvent(RoutedEvent routedEvent, RoutedEventArgs args)
 		{
-			add => RegisterEventHandler(
-				"pointerup",
-				value,
-				eventFilterScript: leftPointerEventFilter,
-				eventExtractorScript: pointerEventExtractor,
-				payloadConverter: PayloadToTappedArgs);
-			remove => UnregisterEventHandler("pointerup", value);
-		}
+			if (_eventHandlerStore.TryGetValue(routedEvent, out List<RoutedEventHandlerInfo> handlers))
+			{
+				foreach (var handler in handlers)
+				{
+					if (!IsHandled(args) || handler.HandledEventsToo)
+					{
+						InvokeHandler(handler.Handler, args);
+					}
+				}
+			}
+
+			var isHandled = IsHandled(args);
+
+			// We don't need to manually bubble up non-handled events on the managed side.
+			// We already take care to bubble up non-handled events on the native side:
+			// - Android: UnoViewGroup.java -> Return false in dispatchTouchEvent
+			// - iOS: UIElement.iOS.cs -> Call base.TouchesBegan, base.TouchesEnded, etc.
+			// - Wasm: WindowManager.ts -> Don't call event.stopPropagation()
+			// In the future, we might decide to stop bubbling up events on the native side, and do it all on the managed side instead.
+			// The first element hit by hit-testing would natively handle the event (stop native bubble up),
+			// become the OriginalSource of the associated RoutedEventArgs, and we would manually bubble up the event on the managed side.
+			var bubblesUpNatively = IsBubblingNatively(args);
+
+			// According to Microsoft, handled events shouldn't bubble up:
+			// https://docs.microsoft.com/en-us/windows/uwp/xaml-platform/events-and-routed-events-overview#the-handled-property
+			// However, an ancestor might have used AddHandler(..., handledEventsToo: true), in which case we need to route the event to it.
+			// Here, we systematically bubble up events in case an ancestor used AddHandler(..., handledEventsToo: true).
+			// This is a naive approach, and it's probably expensive. We should find a better solution.
+			var bubbleUpHandledEventsToo = true;
+
+			if (!bubblesUpNatively && (!isHandled || bubbleUpHandledEventsToo))
+			{
+				// We bubble up the event
+#if __IOS__ || __ANDROID__
+				var parent = this.FindFirstParent<UIElement>();
 #else
-		;
+					var parent = this.GetParent() as UIElement;
 #endif
+				parent?.RaiseEvent(routedEvent, args);
+			}
+		}
 
-#if __WASM__
-		public event KeyEventHandler KeyDown
+		private static bool IsHandled(RoutedEventArgs args)
 		{
-			add => RegisterEventHandler(
-				"keydown",
-				value,
-				eventExtractorScript: "(evt instanceof KeyboardEvent) ? evt.key : \"0\"",
-				payloadConverter: keyStr => new KeyRoutedEventArgs { Key = VirtualKeyHelper.FromKey(keyStr), OriginalSource = this});
-			remove => UnregisterEventHandler("keydown", value);
+			// TODO: WPF reads Handled directly on RoutedEventArgs. 
+			switch (args)
+			{
+				case PointerRoutedEventArgs pointer:
```

Comment:
Can the ICancellableRoutedEventArgs interface be used here?

Reply:
Yep!  But it's a new concept and this code was written before the interface.  Good catch.  Will be refactored in next update.